lib/SIL/SILInstruction.cpp

```diff
@@ -1202,33 +1206,67 @@ SILInstructionResultArray::SILInstructionResultArray(
          "multi-inheritence parent");
   Pointer = reinterpret_cast<const uint8_t *>(Value);
 
+#ifndef NDEBUG
   assert(originalValue == (*this)[0] &&
          "Wrong value returned for single result");
   assert(originalType == (*this)[0]->getType());
 
   auto ValueRange = getValues();
-  (void)ValueRange;
   assert(1 == std::distance(ValueRange.begin(), ValueRange.end()));
   assert(originalValue == *ValueRange.begin());
 
   auto TypedRange = getTypes();
-  (void)TypedRange;
   assert(1 == std::distance(TypedRange.begin(), TypedRange.end()));
   assert(originalType == *TypedRange.begin());
 
   SILInstructionResultArray Copy = *this;
-  (void)Copy;
   assert(Copy.hasSameTypes(*this));
   assert(Copy == *this);
+#endif
+}
+
+SILInstructionResultArray::SILInstructionResultArray(
+    const MultipleValueInstruction *MVI)
+    : Pointer(), Size(1) {
+  // We can perform a straight reinterpret_cast.
+  Pointer = reinterpret_cast<const uint8_t *>(MVI);
+
+#ifndef NDEBUG
+  // Verify our invariants.
+  assert(size() == MVI->Results.size());
+  auto ValueRange = getValues();
+  auto VRangeBegin = ValueRange.begin();
+  auto VRangeIter = VRangeBegin;
+  auto VRangeEnd = ValueRange.end();
+  assert(MVI->Results.size() == std::distance(VRangeBegin, VRangeEnd));
+
+  auto TypedRange = getTypes();
+  auto TRangeBegin = TypedRange.begin();
+  auto TRangeIter = TRangeBegin;
+  auto TRangeEnd = TypedRange.end();
+  assert(MVI->Results.size() == std::distance(VRangeBegin, VRangeEnd));
+  for (unsigned i : indices(MVI->Result)) {
+    assert(MVI->Result[i] == (*this)[i]);
+    assert(MVI->Result[i]->getType() == (*this)[i]->getType());
+    assert(MVI->Result[i] == (*VRangeIter));
+    assert(MVI->Result[i]->getType() == (*VRangeIter)->getType());
+    assert(MVI->Result[i]->getType() == *TRangeIter);
+    ++VRangeIter;
+    ++TRangeIter;
+  }
+
+  SILInstructionResultArray Copy = *this;
+  assert(Copy.hasSameTypes(*this));
+  assert(Copy == *this);
+#endif
 }
 
 SILValue SILInstructionResultArray::operator[](size_t Index) const {
   assert(Index < Size && "Index out of bounds");
-  // Today we only have single instruction results so offset will always be
-  // zero. Once we have multiple instruction results, this will be equal to
-  // sizeof(MultipleValueInstructionResult)*Index. This is safe even to do with
-  // SingleValueInstruction since index will always be zero for the offset.
-  size_t Offset = 0;
+  // *NOTE* In the case where we have a single instruction, Index will always
+  // necessarily be 0 implying that it is safe for us to just multiple Index by
+  // sizeof(MultipleValueInstructionResult).
+  size_t Offset = sizeof(MultipleValueInstructionResult) * Index;
```

Comment:
All this `reinterpret_cast`ing is code smell. It makes more sense to me for `Pointer` to be an opaque pointer type. 
The most self-documenting way to express this is:

```
if (Index == 0)
  return static_cast<ValueBase*>(Pointer);

return static_cast<const MultipleValueInstructionResult*>(Pointer)[Index];
```

If you don't like the branch, you can just speculatively cast from `void *` to `MultipleValueInstructionResult*` with a comment about your cleverness.

If `static_cast` and `reinterpret_cast` are giving you different pointer adjustments, then I think you're already in serious trouble.